TrafficSimulator/Assets/Car/CarSpawner.cs

```diff
@@ -0,0 +1,241 @@
+            {
+                laneLength = 0;
+                _laneNodeTemp = _lanes[j].StartNode;
+
+                // Loop through all lane nodes
+                for (int k = 0; k < _lanes[j].StartNode.Count - 1; k++)
+                {
+                    // Calculate the length of the lane
+                    laneLength = laneLength + Vector3.Distance(_laneNodeTemp.Position, _laneNodeTemp.Next.Position);
+                    _laneNodeTemp = _laneNodeTemp.Next;
+                }
+                _lengths.Add(laneLength);
+            }
+        }
+
+        private void CalculateLaneRatios()
+        {
+            float totalLength = 0;
+
+            // Loop through all lane lengths
+            for (int i = 0; i < _lengths.Count; i++)
+            {
+                totalLength = totalLength + _lengths[i];
+            }
+
+            // Loop through all lane lengths
+            for (int i = 0; i < _lengths.Count; i++)
+            {
+                _ratios.Add(_lengths[i] / totalLength);
+            }
+        }
+
+        private void CalculateLaneIndexes()
+        {
+            int laneIndex;
+
+            // Loop through all roads
+            for (int i = 0; i < _roadSystem.RoadCount; i++)
+            {
+                _roads[i].OnChange();
+                laneIndex = 0;
+
+                // Loop through all lanes
+                for (int j = 0; j < _roads[i].LaneCount; j++)
+                {
+                    _indexes.Add(laneIndex);
+                    laneIndex = laneIndex + 1;
+                }
+            }
+        }
+
+        private void CalculateMaxCarsForLanes()
+        {
+            // Loop through all lanes
+            for (int j = 0; j < _lanes.Count; j++)
+            {
+                // Calculate the max cars for the lane
+                _maxCarsPerLane.Add((int)(_lengths[j] / CarLength));
+            }
+        }
+
+        private void SpawnCarsWithPercentage(float MaxCarsPercentage) 
+        {
+            for (int i = 0; i < _lanes.Count; i++)
+            {
+                // Calculate the number of cars to spawn
+                int carsToSpawn = Mathf.CeilToInt(_maxCarsPerLane[i] * MaxCarsPercentage);
+                if (carsToSpawn == 0)
+                {
+                    return;
+                }
+
+                // Calculate the _offset
+                _offset = _lanes[i].StartNode.Count / carsToSpawn;
+                _laneNodeCurrent = _lanes[i].StartNode;
+
+                // Spawn cars
+                for (int j = 0; j < carsToSpawn; j++)
+                {
+                    // Spawn individual car at current node
+                    SpawnCar(i);
+
+                    _carCounter = _carCounter + 1;
+
+                    SetOffset(_offset);
+                }
+            }
+        }
+
+        private void SpawnCarsWithRatio(int maxCars)
+        {
+            int cars_Spawned = 0;
+            for (int i = 0; i < _lanes.Count; i++)
```

Comment:
The spawn code should be made into its own function as it is used multiple times. As it stands there is code duplication